library/Hasql/Private/Decoders/Row.hs

```diff
@@ -3,14 +3,19 @@ module Hasql.Private.Decoders.Row where
 import Hasql.Private.Prelude hiding (error)
 import Hasql.Private.Errors
 import qualified Database.PostgreSQL.LibPQ as LibPQ
+import qualified Data.Text as T
 import qualified PostgreSQL.Binary.Decoding as A
 import qualified Hasql.Private.Decoders.Value as Value
+import Control.Monad.Fail (MonadFail(..))
 
 
 newtype Row a =
   Row (ReaderT Env (ExceptT RowError IO) a)
   deriving (Functor, Applicative, Monad)
 
+instance MonadFail Row where
+  fail = error . ValueError . T.pack
```

Comment:
This `error` is not the one from `Prelude`! It's the function `error :: RowError -> Row a` defined below in this same module.